# Assignmnet 1

Name - Aryaman Singh Rana 

Roll No. - 1910110093

#### Problem Statement -
To construct a class conditional probability distribution for the three classes present in the Iris data set and find the average error rate.

#### My solution-

#### 1. Importing the necessary files-

In [68]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split 

from sklearn.naive_bayes import GaussianNB

from sklearn.model_selection import GridSearchCV

from sklearn.neighbors import KernelDensity

#### 2. Loading the Iris data set provided -

In [69]:
dataSet = pd.read_csv('iris.data',header=None)
dataSet.head(10)

0    1    2    3            4
0  5.1  3.5  1.4  0.2  Iris-setosa
1  4.9  3.0  1.4  0.2  Iris-setosa
2  4.7  3.2  1.3  0.2  Iris-setosa
3  4.6  3.1  1.5  0.2  Iris-setosa
4  5.0  3.6  1.4  0.2  Iris-setosa
5  5.4  3.9  1.7  0.4  Iris-setosa
6  4.6  3.4  1.4  0.3  Iris-setosa
7  5.0  3.4  1.5  0.2  Iris-setosa
8  4.4  2.9  1.4  0.2  Iris-setosa
9  4.9  3.1  1.5  0.1  Iris-setosa

#### 2.1 Checking the data set provided -

In [70]:
 dataSet.groupby(4).size()

4
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
dtype: int64

We see that the given 3 class are equal in number.

---

#### 3.0 Seperating the data into independent varible and dependent variables

In [71]:
# X being the dependent variable and y being the independent variable
X = dataSet.iloc[:,:-1].values
y = dataSet.iloc[:,-1].values

#### 3.1 Spliting the data into training set and test set -

|S No.|Training set|Test set|
|---|---|---
|1|80%|20%|
|2|70%|30%|
|3|60%|40%|

In [72]:
X_train1, X_test1, y_train1, y_test1 = train_test_split(X,y, train_size = 0.8, test_size = 0.2)

X_train2, X_test2, y_train2, y_test2 = train_test_split(X,y, train_size = 0.7, test_size = 0.3)

X_train3, X_test3, y_train3, y_test3 = train_test_split(X,y, train_size = 0.6, test_size = 0.4)

#### 4.0 

Using GridSearchCV to find the bandwidth and the kernel type

In [73]:
parameters = {'kernel':['gaussian','tophat','epanechnikov','exponential','linear','cosine'],'bandwidth' :np.logspace(-1,1,20)}

# for ratio 80%-20%
gsVar1 = GridSearchCV(KernelDensity(),parameters)
gsVar1.fit(X_train1)
dictionary1 = gsVar1.best_params_
print("-------------------------")

# for ratio 70%-30%
gsVar2 = GridSearchCV(KernelDensity(),parameters)
gsVar2.fit(X_train2)
dictionary2 = gsVar2.best_params_
print("-------------------------")

# for ratio 60%-40%
gsVar3 = GridSearchCV(KernelDensity(),parameters)
gsVar3.fit(X_train3)
dictionary3 = gsVar3.best_params_
print("-------------------------")

print(dictionary1)
print(dictionary2)
print(dictionary3)

C:\Users\wickedbaba\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:922: UserWarning: One or more of the test scores are non-finite: [ -87.27704334          -inf          -inf  -45.62634809          -inf
           nan  -60.22693687          -inf          -inf  -49.27254821
          -inf           nan  -49.12397756          -inf          -inf
  -55.46317837          -inf           nan  -47.21671538          -inf
          -inf  -63.7600421           -inf           nan  -50.85228597
          -inf          -inf  -73.8728822           -inf           nan
  -58.22404781          -inf          -inf  -85.58774128          -inf
           nan  -68.36951668          -inf          -inf  -98.72641865
          -inf           nan  -80.7518844           -inf          -inf
 -113.12149672          -inf           nan  -95.05741442          -inf
          -inf -128.59906205          -inf           nan -110.9958183
  -68.40906889  -60.38436548 -144.97889234  -58.7323908            nan
 

-------------------------


C:\Users\wickedbaba\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:929: RuntimeWarning: invalid value encountered in subtract
  array_stds = np.sqrt(np.average((array -


-------------------------
-------------------------
{'bandwidth': 0.1, 'kernel': 'exponential'}
{'bandwidth': 0.20691380811147897, 'kernel': 'gaussian'}
{'bandwidth': 0.1, 'kernel': 'exponential'}


C:\Users\wickedbaba\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:922: UserWarning: One or more of the test scores are non-finite: [ -79.50959782          -inf          -inf  -37.48552129          -inf
           nan  -52.77195971          -inf          -inf  -39.45985268
          -inf           nan  -40.8987188           -inf          -inf
  -43.6667753           -inf           nan  -37.8389009           -inf
          -inf  -49.62103079          -inf           nan  -40.02213139
          -inf          -inf  -57.00266973          -inf           nan
  -45.3970071           -inf          -inf  -65.61049533          -inf
           nan  -52.87341864          -inf          -inf  -75.30420201
          -inf           nan  -61.96032032          -inf          -inf
  -85.95793995          -inf           nan  -72.47959288          -inf
          -inf  -97.43773247          -inf           nan  -84.26419753
          -inf          -inf -109.60432291          -inf           nan


#### 4.1 I take the given bandwidth and kernel value and use it to find the Kernel density on the train dataset

In [74]:
labelArray = ['Iris-setosa','Iris-versicolor','Iris-virginica']
kdArray1 = []
kdArray2 = []
kdArray3 = []

for element in labelArray:
    # for ratio 1
    kd1 = KernelDensity(kernel = dictionary1['kernel'],bandwidth = dictionary1['bandwidth']).fit(X_train1[y_train1 == element])
    kdArray1.append(kd1)
    
    # for ratio 2
    kd2 = KernelDensity(kernel = dictionary2['kernel'],bandwidth = dictionary2['bandwidth']).fit(X_train2[y_train2 == element])
    kdArray2.append(kd2)
    
    # for ratio 3
    kd3 = KernelDensity(kernel = dictionary3['kernel'],bandwidth = dictionary3['bandwidth']).fit(X_train3[y_train3 == element])
    kdArray3.append(kd3)

# checking the score samples
for element in kdArray1:
    print(element.score_samples([X_train1[0]]))

print('------------------')

for element in kdArray2:
    print(element.score_samples([X_train2[0]]))

print('------------------')

for element in kdArray3:
    print(element.score_samples([X_train3[0]]))

[-48.80371994]
[-12.35349601]
[0.81782054]
------------------
[-209.74427242]
[-9.69900527]
[-0.27324078]
------------------
[-29.24049822]
[-5.46324114]
[1.00217083]


#### 4.2 Now I make a prediction matrix which will be used to compare with the testing data set 

In [75]:
# for ratio 80%-20%
prediction1 = []

for i in range(len(X_test1)):
    setosa = kdArray1[0].score_samples([X_test1[i]])
    versi =  kdArray1[1].score_samples([X_test1[i]]) 
    virgi =  kdArray1[2].score_samples([X_test1[i]]) 
    
    if(setosa > versi):
        if(setosa > virgi):
            prediction1.append(labelArray[0])
            
        else:
            prediction1.append(labelArray[2])
    else:
        if(versi>virgi):
            prediction1.append(labelArray[1])
        else:
            prediction1.append(labelArray[2])
            

# for ratio 70%-30%
prediction2 = []

for i in range(len(X_test2)):
    setosa =  kdArray2[0].score_samples([X_test2[i]])
    versi = kdArray2[1].score_samples([X_test2[i]]) 
    virgi = kdArray2[2].score_samples([X_test2[i]]) 
    
    if(setosa > versi):
        if(setosa > virgi):
            prediction2.append(labelArray[0])
            
        else:
            prediction2.append(labelArray[2])
    else:
        if(versi > virgi):
            prediction2.append(labelArray[1])
        else:
            prediction2.append(labelArray[2])
            
            
# for ratio 60%-40%
prediction3 = []

for i in range(len(X_test3)):
    setosa = kdArray3[0].score_samples([X_test3[i]])
    versi = kdArray3[1].score_samples([X_test3[i]]) 
    virgi = kdArray3[2].score_samples([X_test3[i]]) 
    
    if(setosa > versi):
        if(setosa > virgi):
            prediction3.append(labelArray[0])
            
        else:
            prediction3.append(labelArray[2])
    else:
        if(versi > virgi):
            prediction3.append(labelArray[1])
        else:
            prediction3.append(labelArray[2])
            
print(len(prediction1))
print(len(prediction2))
print(len(prediction3))

30
45
60


The length confirms that the program is outputting the right ammount of labels required per every ratio fed to it.

#### 4.3 Now I compare these prediciton matrix with y_test to find the error rate

In [76]:
negative1 = 0
negative2 = 0
negative3 = 0
testLen1 = len(y_test1)
testLen2 = len(y_test2)
testLen3 = len(y_test3)

for i in range(len(y_test1)):
    if(prediction1[i] == y_test1[i]):
        negative1+=1

for i in range(len(y_test2)):
    if(prediction2[i] == y_test2[i]):
        negative2+=1

for i in range(len(y_test3)):
    if(prediction3[i] == y_test3[i]):
        negative3+=1
        

error1 = ((testLen1-negative1)/testLen1)*100
print("The error of ratio 80%-20% - ",error1 ,"%")

error2 = ((testLen2-negative2)/testLen2)*100
print("The error of ratio 70%-30% - ",error2 ,"%")

error3 = ((testLen3-negative3)/testLen3)*100
print("The error of ratio 60%-40% - ",error3 ,"%")

print()
average = (error1 + error2 + error3 )/3
print("Average error",average,"%")

The error of ratio 80%-20% -  0.0 %
The error of ratio 70%-30% -  6.666666666666667 %
The error of ratio 60%-40% -  3.3333333333333335 %

Average error 3.3333333333333335 %


#### 4.4 Now, I ran these test 10 times with different seed values and recorded the follwoing error values -

|Test No.|Error in Ratios -|80%-20%|70%-30%|60%-40%|Average -|
|---|---|---|---|---|---|
|1||3.33%|2.22%|5.0%|3.518%|
|2||3.33%|2.22%|1.66%|2.407%|
|3||3.33%|2.22%|5.0%|3.518%|
|4||0.0%|4.44%|5.0%|3.148%|
|5||3.33%|0.0%|3.33%|2.22%|
|6||3.33%|4.44%|5.0%|4.259%|
|7||3.33%|2.22%|3.33%|2.96%|
|8||6.66%|4.44%|1.66%|4.259%|
|9||3.33%|4.44%|5.0%|4.25%|
|10||3.33%|2.22%|1.66%|2.40%|


#### 5.0  Taking the average error -

In [77]:
avg = (3.5 + 2.4+3.5+3.1+2.2+4.2+2.9+4.2+4.2+2.4)/10

print("Net AVG Error = ",avg,"%" )

Net AVG Error =  3.2599999999999993 %


#### In online sources the average error rate comes out to be 2%-3% and the aforementioned average error is near to it.

#### 6.0 Conclusion -

1. With the given data set size of 150 entries with 4 independent variable and 1 independent variable, the whole program ran in under 3 sec, giving an average error rate of 3.25%.  This is given the fact that I have an 8Gb RAM and an Intel i7 processor. 


2. When we go on to increase the size of the data set, not only does its time complexity increases but also the space complexity increases.<br>

     _Take the example of_ **train_test_split** - The actual time complexity of **train_test_split** is O(n). But, if the data set is large, the machine has to access a lot of memory locations, which creates a bottleneck based on data passing and retrieval. Usually, the machine can do 10^8 operations per sec. But this problem alone results in the program taking longer than usual. 
     

3. Now, on the basis of above written program, **net time complexity** of this **BDR classifier** would be
    **O(n*f)**. where f = number of features and n = dataset size.

4. When we increase the number of attributes and feature vectors in the learning set, the time complexity of the BDR classifier **increases** - (as TC = O(n*f) and we are increasing the value of f) 